### PR TITLE
[Gradient Compression] Add a minimum compression rate threshold for PowerSGD communication hook

### PR DESCRIPTION
**Pull request by lazysjb**
Fixes #{52034}
- Add a minimum compression rate threshold to `PowerSGDState`
- Use the threshold to determine whether to compress high-rank tensors or not

### CODE REVIEW

**Comment by lazysjb** [comment]
fyi, the relevant task id is T84998280

**Review by wayi1 — APPROVED** [review]

**Comment by SciPioneer** [comment]
@tvogels Could you please take a look?

If compressing a tensor cannot yield enough bandwidth saving (at least 50% by default), then this high-rank tensor will be moved to rank-1 tensor list.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -247,29 +266,43 @@ def powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
...
-    # Step I: Handle rank-1 tensors.
+    # Step I: Divide all the tensors into two groups,
+    # one will be compressed before allreduce and the other will be directly allreduced without compression.
+    rank1_tensors, high_rank_tensors, high_rank_tensors_to_compress = [], [], []
+    for tensor in tensors:
+        if tensor.ndimension() <= 1:
+            rank1_tensors.append(tensor)
+        else:
+            high_rank_tensors.append(tensor.view(tensor.shape[0], -1))
+
```
Comment (tvogels) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:277]: I think the logic can be made a bit more elegant/concise by including the 'rank 1' case in `_should_compress`.

`_should_compress` could look like:
```python
n, m = tensor.shape[0], tensor.nelement() // tensor.shape[0]
# same code as now
```

Rank-1 matrices would be 'n x 1', and compressing them would always be bad according to the logic of `min_compression_rate`.

The main logic could then remain simpler:

```
compressed_tensors = [tensor for tensor in tensors if _should_compress(tensor, ...)]
uncompressed_tensors = [tensor for tensor in tensors if not _should_compress(tensor, ...)]
```

**Review by tvogels — CHANGES_REQUESTED** [review]
Looks good! I have one suggestion that aims to keep the logic leaner.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 1 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:277]: Thanks for the suggestion! This structure will look better, as we don't really need a hard boundary between rank-1 and high-rank tensors, only need a boundary between compressed and uncompressed tensors.

I will create a follow-up PR to address this issue.

Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:277]: Resolved in #52981

**Pull request closed without merging**